test/e2e/es/certs_test.go

```diff
@@ -32,24 +32,133 @@ import (
+						// let's clean up the CA secret here.
+						toDelete := mkCertSecret(nil, nil)
+						_ = k.Client.Delete(context.Background(), &toDelete)
+					},
+				},
+			}
+		},
+	}
+
+	// The initial Cluster builder should result in a healthy cluster as verified by the standard check steps
+	// Now modify cluster to use custom HTTP CA but simulate a user error by populating the secret with
+	// garbage and verify this is bubbled up through an event
+	b := initialCluster.
+		WithCustomHTTPCerts(caSecretName).
+		WithMutatedFrom(&initialCluster)
+
+	withBogusCert := test.WrappedBuilder{
+		BuildingThis: b,
+
+		// we don't want to run a regular mutation as the invalid certificates will not lead to a successful upgrade
+		OverrideMutationSteps: func(k *test.K8sClient) test.StepList {
+			steps := test.StepList{
+				{
+					Name: "Create an invalid CA secret",
+					Test: test.Eventually(func() error {
+						bogusSecret := mkCertSecret([]byte("garbage"), []byte("more garbage"))
+						_, err := reconciler.ReconcileSecret(k.Client, bogusSecret, nil)
+						return err
+					}),
+				},
+			}
+			steps = append(steps, b.UpgradeTestSteps(k)...)
+			steps = append(steps,
+				test.Step{
+					Name: "Invalid CA secret should create events",
+					Test: test.Eventually(func() error {
+						eventList, err := k.GetEvents(test.EventListOptions(esNamespace, initialCluster.Elasticsearch.Name)...)
+						if err != nil {
+							return err
+						}
+						for _, evt := range eventList {
+							if evt.Type == corev1.EventTypeWarning &&
+								evt.Reason == events.EventReconciliationError &&
+								strings.Contains(evt.Message, "can't parse") {
+								return nil
+							}
+						}
+						return fmt.Errorf("expected validation event but could not observe it")
+					}),
+				},
+			)
+			return steps
+		},
+	}
+
+	var ca *certificates.CA
+
+	// A NOOP mutation but set up the CA secret correctly now (using the existing CA generation code in the operator)
+	withCustomCert := test.WrappedBuilder{
+		BuildingThis: initialCluster.
+			WithCustomHTTPCerts(caSecretName).
+			WithMutatedFrom(&initialCluster),
+
+		PreMutationSteps: func(k *test.K8sClient) test.StepList {
+			return test.StepList{
+				{
+					Name: "Create custom CA secret",
+					Test: test.Eventually(func() error {
+						var err error
+						ca, err = certificates.NewSelfSignedCA(certificates.CABuilderOptions{
+							Subject: pkix.Name{
+								CommonName:         "eck-e2e-test-custom-ca",
+								OrganizationalUnit: []string{"eck-e2e"},
+							},
+						})
+						if err != nil {
+							return err
+						}
+
+						caSecret := mkCertSecret(
+							certificates.EncodePEMCert(ca.Cert.Raw),
+							certificates.EncodePEMPrivateKey(*ca.PrivateKey),
+						)
+						_, err = reconciler.ReconcileSecret(k.Client, caSecret, nil)
+						return err
+					}),
+				},
+			}
+		},
+		// verification that the certificate works is implicit by the standard check steps that all require an Elasticsearch
```

Comment:
How do we know from here that the operator has reconciled the cluster and the updated custom Secret is in use ?

Reply:
I added a new check step called `CheckHTTPConnectivity` which should create a test request to each Pod. But you are right that there is a gap in the logic in that this step relies on the `*-es-certs-public` secret being reconciled and then uses the CA from that secret to make the connections.

Reply:
I think could add an explicit step that verifies with the custom CA.

Reply:
Also if I slow down the operator and delay the reconciliation I can see this kind of error:

```
    cluster_uuid.go:57: 
        	Error Trace:	cluster_uuid.go:57
        	Error:      	Received unexpected error:
        	            	Get "https://test-custom-http-ca-2fxk-es-http.e2e-mercury.svc:9200/": x509: certificate signed by unknown authority
        	Test:       	TestCustomHTTPCA/Cluster_UUID_should_have_been_preserved#01
```

Not sure if it is a realistic experiment though...

Reply:
Slowing down means you also delayed the certificate secret reconciliation to run into the condition I [mentioned](https://github.com/elastic/cloud-on-k8s/pull/4575#discussion_r659671813)?

Reply:
Yes, I think `CheckHTTPConnectivity` was run before the certificate was reconciled:
 
![image](https://user-images.githubusercontent.com/976373/123630876-5a40ac00-d816-11eb-894b-0a76fd48329f.png)

You're right, it is just another manifestation of the same condition. Sorry, I'm just trying to understand if this test could become a flaky one once deployed 🙃

Reply:
No need to be sorry. Very grateful for your thorough review!

Reply:
I just checked out last commit ([5e29d98](https://github.com/elastic/cloud-on-k8s/pull/4575/commits/5e29d98ac5c11745d9ccb5649a9e08f54baa1ac6)), it looks good but I think we need a similar step to validate that at step `reconfigure back to self-signed operator provided CA` the custom CA is no more in use.